lib/nmatrix/lapack.rb

```diff
@@ -106,9 +165,42 @@ def self.qr(matrix, mode: "full", pivoting: false)
     end
     m, n = matrix.shape
 
-    if pivoting == false
+    if pivoting == true
+      qr, tau, jpvt = NumRuby::Lapack.geqp3(matrix)
+      jpvt -= 1
+    else
       qr, tau = NumRuby::Lapack.geqrf(matrix)
+    end
+
+    # calculate R here for both pivot true & false
+
+    if ['economic', 'raw'].include?(mode.downcase) or m < n
+      r = NumRuby.triu(matrix)
+    else
+      r = NumRuby.triu(matrix[0...n, 0...n])
+    end
+
+    if pivoting == true
```

Comment:
This validation is similar to line `168`. I think you could put this in a function. WDYT? 🤔

Reply:
Agreed. cc @Uditgulati

Reply:
A function for an if statement?

Reply:
🤔 yes, maybe move only the conditional is weird, but I think that you could the conditional blocks in functions. WDYT? 🤔